File: src/clock.py
import datetime
import os
import typing

import pygame
import pytz

pygame.font.init()


class Clock:
    timezone: typing.Optional[pytz.BaseTzInfo]
    am_pm: bool  # Poiuu You are Fucking Shit
    name: str
    date_font = pygame.font.Font(os.getcwd() + "/src/fonts/digital.ttf", 40)
    normal_font = pygame.font.Font(pygame.font.get_default_font(), 20)
    revert: bool

    def __init__(self, config: str, am_pm: bool, revert: bool) -> None:
        if config.lower() != "local":
            self.timezone = pytz.timezone(config)
            self.name = config
        else:
            self.timezone = None
            self.name = "Local time"
        self.am_pm = am_pm
        self.revert = revert

    def convert_time_to_tz(self) -> datetime.datetime:
        now = datetime.datetime.now(self.timezone)
        return now

    def render(
        self, time: datetime.datetime, surface: pygame.Surface
    ) -> pygame.Surface:
        if not self.am_pm:
            timer = time.strftime("%H:%M:%S")
        else:
            timer = time.strftime("%I:%M:%S %p")
        middle_y = self.__get_middle(surface)
        self.__center(timer, surface, self.date_font, middle_y)
        if self.timezone:
            offset = self.name
        else:
            offset = f"local time."
        zone_thing = time.strftime("%z")
        z_negative_or_positive = zone_thing[:1]
        z_hour = zone_thing[1:3]
        z_minute = zone_thing[3:5]
        if z_negative_or_positive == "-":
            zone = f"{z_hour}:{z_minute} hours behind"
        elif z_negative_or_positive == "+":
            zone = f"{z_hour}:{z_minute} hours forward"
        else:
            zone = "Local time"
        try:
            if int(time.strftime("%Z")) < 0 or int(time.strftime("%Z")) > 0:
                date_thing = f"{time.strftime('%A %d/%B/%Y UTC%Z')} ({zone})"
        except ValueError:
            date_thing = f"{time.strftime('%A %d/%B/%Y %Z')} ({zone})"

        offset_middle_y = middle_y + 90
        self.__center(f"Currently {offset}", surface, self.normal_font, offset_middle_y)
        self.__center(f"The date is {date_thing}", surface, self.normal_font, offset_middle_y + 40)  # type: ignore
        return surface

    def __get_middle(self, surface: pygame.Surface) -> int:
        _, h = surface.get_size()
        return h // 2

    def __center(
        self,
        text: str,
        window: pygame.Surface,
        font: pygame.font.Font,
        y: typing.Optional[int],
    ):
        max_width, _ = window.get_size()
        wrapped_lines = self.__word_wrap(text, max_width, font)

        total_height = len(wrapped_lines) * font.get_height()
        y_centered = y - total_height // 2  # type: ignore

        for line in wrapped_lines:
            rendered = self.__render_font(font, line)
            rect = rendered.get_rect()
            rect.topleft = (
                max_width - rect.width
            ) // 2, y_centered  # Center horizontally
            self.__to_screen(rendered, window, rect=rect)
            y_centered += font.get_height()

    def __render_font(self, font: pygame.font.Font, text: str) -> pygame.Surface:
        return font.render(text, True, "white" if not self.revert else "black")

    def __to_screen(
        self,
        text: pygame.Surface,
        window: pygame.Surface,
        dest: typing.Optional[tuple[int, int]] = None,
        rect: typing.Optional[pygame.Rect] = None,
    ) -> None:
        window.blit(text, rect if rect is not None else dest)  # type: ignore

    def __word_wrap(self, text: str, max_width: int, font: pygame.font.Font):
        words = text.split(" ")
        lines = []
        current_line = ""

        for word in words:
            test_line = current_line + word + " "
            test_width, _ = font.size(test_line)

            if test_width <= max_width:
                current_line = test_line
            else:
                lines.append(current_line.rstrip())
                current_line = word + " "

        lines.append(current_line.rstrip())

        return lines

File: src/clock.rs
use crate::snow;
use chrono::{self, Datelike, TimeZone};
use sdl2;
use slicestring::Slice;
// use humantime;
// use crate::snow;

pub struct Clock<'a, 'b, 'c, 'd> {
    timezone: Option<chrono_tz::Tz>,
    am_pm: bool,
    name: String,
    revert: bool,
    date_font: &'c sdl2::ttf::Font<'a, 'b>,
    normal_font: &'c sdl2::ttf::Font<'a, 'b>,
    markery: std::marker::PhantomData<&'d str>, // used for 'd for the surface
                                                // snow: snow::SnowParticles,
}

impl<'a, 'b, 'c, 'd> Clock<'a, 'b, 'c, 'd> {
    pub fn new(
        timezone: Option<chrono_tz::Tz>,
        am_pm: bool,
        revert: bool,
        date_font: &'c sdl2::ttf::Font<'a, 'b>,
        normal_font: &'c sdl2::ttf::Font<'a, 'b>,
    ) -> Self {
        let name: String;
        if timezone.is_none() {
            name = "Local".to_owned();
        } else {
            name = timezone.unwrap().name().to_owned();
        }

        return Self {
            timezone,
            am_pm,
            name,
            revert,
            date_font,
            normal_font,
            markery: std::marker::PhantomData,
        };
    }

    pub fn current_datetime_in_timezone(&self) -> chrono::DateTime<chrono::FixedOffset> {
        match self.timezone {
            Some(tz) => {
                // println!("It has timezone");
                tz.from_utc_datetime(&chrono::Utc::now().naive_utc())
                    // .unwrap()
                    .fixed_offset()
            }
            None => {
                // println!("It's none");
                chrono::Local::now().fixed_offset()
            }
        }
    }

    pub fn render(
        &self,
        time: chrono::DateTime<chrono::FixedOffset>,
        surface: &mut sdl2::surface::Surface,
        snow_particle: &mut snow::SnowParticles,
    ) {
        let utcized = time.naive_local();
        let new_year = chrono::NaiveDate::from_ymd_opt(utcized.year() + 1, 1, 1).unwrap();
        let time_new_year = chrono::NaiveTime::from_hms_opt(0, 0, 0).unwrap();
        let built = chrono::NaiveDateTime::new(new_year, time_new_year);

        let timer = {
            if !self.am_pm {
                time.format("%H:%M:%S").to_string()
            } else {
                time.format("%I:%M:%S %p").to_string()
            }
        };
        let middle_y = {
            let (_, h) = surface.size();
            (h / 2) as i32
        };
        self.center(&timer, surface, self.date_font, Some(middle_y - 100));
        let offset: &str;
        if self.timezone.is_some() {
            offset = &self.name
        } else {
            offset = "Local time"
        }
        let zt = &time.format("%z").to_string();
        let z_n_or_p = zt.slice(..1);
        let zh = zt.slice(1..3);
        let zm = zt.slice(3..5);
        let zone = {
            if z_n_or_p == "-" {
                format!("{}:{} hours behind", zh, zm)
            } else if z_n_or_p == "+" {
                format!("{}:{} hours forward", zh, zm)
            } else {
                "Local time".to_owned()
            }
        };
        let d_thing = match time.format("%Z").to_string().parse::<i32>() {
            Ok(_) => {
                format!(
                    "{} ({})",
                    time.format("%A %d/%B/%Y UTC%Z").to_string(),
                    zone
                )
            }
            Err(_) => {
                format!("{} {}", time.format("%A %d/%B/%Y %Z").to_string(), zone)
            }
        };
        self.center(
            format!("Currently {}", offset).as_str(),
            surface,
            self.normal_font,
            Some(middle_y + 20),
        );
        self.center(
            format!("The date is {}", d_thing).as_str(),
            surface,
            self.normal_font,
            Some(middle_y + 70),
        );
        let left = built - utcized;

        // let formatted = humantime::format_duration(left.to_std().unwrap()).to_string();
        let formatted = self.formatter(left);

        self.center(
            formatted.as_str(),
            surface,
            self.normal_font,
            Some(middle_y + 120),
        );

        snow_particle.render(surface, self.revert);
    }

    fn formatter(&self, left: chrono::Duration) -> String {
        let stded = left.to_std().unwrap();
        let seconds = stded.as_secs() as i64;
        let nanoseconds = stded.subsec_millis();

        let years = seconds / (365 * 24 * 60 * 60);
        let remaining_seconds = seconds % (365 * 24 * 60 * 60);

        let months = remaining_seconds / (30 * 24 * 60 * 60);
        let remaining_seconds = remaining_seconds % (30 * 24 * 60 * 60);

        let weeks = remaining_seconds / (7 * 24 * 60 * 60);
        let remaining_seconds = remaining_seconds % (7 * 24 * 60 * 60);

        let days = remaining_seconds / (24 * 60 * 60);
        let remaining_seconds = remaining_seconds % (24 * 60 * 60);

        let hours = remaining_seconds / 3600;
        let remaining_seconds = remaining_seconds % 3600;

        let minutes = remaining_seconds / 60;
        let remaining_seconds = remaining_seconds % 60;

        let mut build = String::new();
        build += "New years: ";

        if left.num_hours() <= 12 {
            // new year 12 hours left
            if years != 0 {
                build += format!("{} years, ", years).as_str();
            }
            if months != 0 {
                build += format!("{} months, ", months).as_str();
            }
            if weeks != 0 {
                build += format!("{} weeks, ", weeks).as_str();
            }
            if days != 0 {
                build += format!("{} days, ", days).as_str();
            }
            build += format!(
                "{} hours, {} minutes, {} seconds, {} miliseconds",
                hours, minutes, remaining_seconds, nanoseconds
            )
            .as_str();
            return build;
        }
        return "".to_string();
    }

    fn center(
        &self,
        text: &str,
        surface: &mut sdl2::surface::Surface,
        font: &'c sdl2::ttf::Font<'a, 'b>,
        y: Option<i32>,
    ) {
        let y = match y {
            Some(s) => s,
            None => 0,
        };
        let (mw, _) = surface.size();
        let wrapped = self.word_wrap(text, mw, font);

        let h = wrapped.len() as i32 * font.height();
        let mut y_centered = (y - h / 2) as i32;

        for line in wrapped {
            let rendered = match self.render_font(font, &line) {
                Ok(s) => s,
                Err(_) => continue,
            };
            // let rect = sdl2::rect::Rect::new(rendered.width() as i32, rendered.height() as i32, (surface.width() - rendered.width()) / 2, y_centered as u32);
            let rect = sdl2::rect::Rect::new(
                ((surface.width() - rendered.width()) / 2) as i32,
                y_centered as i32,
                rendered.width(),
                rendered.height(),
            );
            self.to_screen(&rendered, surface, None, Some(rect));
            y_centered += font.height();
        }
    }
    fn render_font(
        &self,
        font: &'c sdl2::ttf::Font<'a, 'b>,
        text: &str,
    ) -> sdl2::ttf::FontResult<sdl2::surface::Surface<'d>> {
        font.render(text).blended({
            if self.revert {
                sdl2::pixels::Color::BLACK
            } else {
                sdl2::pixels::Color::WHITE
            }
        })
    }

    fn to_screen(
        &self,
        surface: &sdl2::surface::Surface,
        mut target: &mut sdl2::surface::Surface,
        dest: Option<(i32, i32)>,
        rect: Option<sdl2::rect::Rect>,
    ) {
        surface
            .blit(None, &mut target, {
                match (dest, rect) {
                    (Some(_), None) => {
                        let (w, h) = surface.size();
                        sdl2::rect::Rect::new(0, 0, w, h)
                    }
                    (None, Some(r)) => r,
                    _ => {
                        panic!("Unexpected!");
                    }
                }
            })
            .unwrap()
            .unwrap();
    }

    fn word_wrap(
        &self,
        text: &str,
        max_width: u32,
        font: &'c sdl2::ttf::Font<'a, 'b>,
    ) -> Vec<String> {
        let words = text.split_whitespace();
        let mut lines: Vec<String> = Vec::new();
        let mut current_line = String::new();

        for word in words {
            let test_line = current_line.clone() + word + " ";
            let (test_width, _) = font.size_of(&test_line).unwrap();

            if test_width <= max_width {
                current_line = test_line;
            } else {
                lines.push(current_line.trim_end().to_string());
                current_line = word.to_owned() + " ";
            }
        }

        if !current_line.trim().is_empty() {
            lines.push(current_line.trim_end().to_string());
        }

        // println!("{:#?}", lines);
        lines
    }
}
